File: serve-spa.mjs
#!/usr/bin/env node
/**
 * Simple SPA server for local testing
 * Serves dist files, falls back to index.html for routes
 */
import http from 'http';
import fs from 'fs';
import path from 'path';
import url from 'url';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const PORT = 8080;
const DIST_DIR = path.join(__dirname, 'dist');

const server = http.createServer((req, res) => {
  const parsedUrl = url.parse(req.url, true);
  let pathname = parsedUrl.pathname || '/';

  try {
    pathname = decodeURIComponent(pathname);
  } catch {
    pathname = '/';
  }

  // Remove trailing slash except for root
  if (pathname !== '/' && pathname.endsWith('/')) {
    pathname = pathname.slice(0, -1);
  }

  const safePath = pathname.replace(/^\.\.(?:\/|$)/, '/');
  let filePath = path.join(DIST_DIR, safePath);

  // Check if it's a file
  if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) {
    const ext = path.extname(filePath);
    const mimeTypes = {
      '.html': 'text/html',
      '.js': 'application/javascript',
      '.css': 'text/css',
      '.json': 'application/json',
      '.png': 'image/png',
      '.jpg': 'image/jpeg',
      '.jpeg': 'image/jpeg',
      '.gif': 'image/gif',
      '.svg': 'image/svg+xml',
      '.woff': 'font/woff',
      '.woff2': 'font/woff2',
      '.ttf': 'font/ttf',
      '.eot': 'application/vnd.ms-fontobject',
    };

    res.writeHead(200, { 'Content-Type': mimeTypes[ext] || 'text/plain' });
    res.end(fs.readFileSync(filePath));
    return;
  }

  // Check if it's a directory with index.html
  if (fs.existsSync(filePath) && fs.statSync(filePath).isDirectory()) {
    const indexPath = path.join(filePath, 'index.html');
    if (fs.existsSync(indexPath)) {
      res.writeHead(200, { 'Content-Type': 'text/html' });
      res.end(fs.readFileSync(indexPath));
      return;
    }
  }

  // SPA fallback: serve index.html for all other routes
  const indexPath = path.join(DIST_DIR, 'index.html');
  if (fs.existsSync(indexPath)) {
    res.writeHead(200, { 'Content-Type': 'text/html' });
    res.end(fs.readFileSync(indexPath));
    return;
  }

  // Not found
  res.writeHead(404, { 'Content-Type': 'text/html' });
  res.end('<h1>404 Not Found</h1><p>dist/index.html not found. Run npm run build first.</p>');
});

server.listen(PORT, () => {
  console.log(`🚀 SPA server running at http://localhost:${PORT}`);
  console.log(`📁 Serving files from: ${DIST_DIR}`);
  console.log(`\n✅ Test routes:`);
  console.log(`   • http://localhost:${PORT}/ (Landing)`);
  console.log(`   • http://localhost:${PORT}/staff (Staff Dashboard - should show page, not 404)`);
  console.log(`   • http://localhost:${PORT}/resident (Resident Dashboard)`);
  console.log(`   • http://localhost:${PORT}/admin (Admin Dashboard)\n`);
  console.log(`Press Ctrl+C to stop\n`);
});

File: serve-spa.py
#!/usr/bin/env python3
"""
Simple HTTP server for testing SPA routing locally.
Mimics nginx behavior: serves dist files, falls back to index.html for routes.
"""
import http.server
import socketserver
import os
from pathlib import Path
from urllib.parse import unquote, urlparse

PORT = 8080
DIST_DIR = Path(__file__).parent / "dist"

class SPARequestHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(DIST_DIR), **kwargs)

    def do_GET(self):
        # Parse the request path
        parsed_path = urlparse(self.path)
        safe_path = parsed_path.path or '/'

        try:
            safe_path = unquote(safe_path)
        except Exception:
            safe_path = '/'

        file_path = DIST_DIR / safe_path.lstrip('/')
        
        # If it's a request for a file that exists, serve it
        if file_path.is_file():
            return super().do_GET()
        
        # If it's a directory request, try index.html
        if file_path.is_dir() and (file_path / 'index.html').is_file():
            self.path = f"{parsed_path.path}/index.html"
            return super().do_GET()
        
        # Otherwise, serve index.html (SPA route)
        self.path = '/index.html'
        return super().do_GET()

if __name__ == '__main__':
    if not DIST_DIR.exists():
        print(f"Error: {DIST_DIR} does not exist. Run 'npm run build' first.")
        exit(1)
    
    handler = SPARequestHandler
    with socketserver.TCPServer(("", PORT), handler) as httpd:
        print(f"🚀 SPA server running at http://localhost:{PORT}")
        print(f"📁 Serving files from: {DIST_DIR}")
        print(f"✅ Navigate to http://localhost:{PORT}/staff to test the route")
        print(f"   Press Ctrl+C to stop\n")
        httpd.serve_forever()
